### PR TITLE
Add container MVP and update webhook to support materialize root resource creation

### PR DESCRIPTION
**Pull request by LeiQL**
- **Container MVP**: add container MVP to support basic CRUD calls.
- **Materialize**: create container resources for versioned objects to support resource sync-back in materialize as container resource is required when version resource needs to sync back ARM.
- **Materialize**: require materialized obj `metadata.name` formatted `"containerName:version"` in the properties of a catalog to ensure `managers` know what container resource to be created. And this materialized resource using `"containerName-version"` as metadata name. For the catalog spec sample below , `webappconfig` as container resource name and `webappconfig-v1` as materialized resource name.
````
properties:
    metadata: 
      name: webappconfig:v1
````
- **Webhook**:  check and handle `owner reference` creation if container resource exists in mutation webhook
- **Webhook**:  check object name should consist of `<rootResource>-<version>` in validation webhook,
- **Webhook**:  check if `spec.RootResource` exists and `ownerReference` should not be empty list in validation webhook,

**TODO:**
- Update sample/docs based on versioning experience/reference rule

### CODE REVIEW

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go [hunk 1]
```diff
@@ -179,11 +205,36 @@ func (i *MaterializeStageProvider) Process(ctx context.Context, mgrContext conte
...
+						solutionState.Spec.RootResource = parts[0]
+						solutionState.Spec.Version = parts[1]
+					} else {
+						mLog.Errorf("Solution name is invalid: solution - %s, catalog - %s", solutionName, name)
+						return outputs, false, v1alpha2.NewCOAError(nil, fmt.Sprintf("Invalid solution name: catalog - %s", name), v1alpha2.BadRequest)
+					}
+
+					mLog.Debugf("  P (Materialize Processor): check solution contains %v, namespace %s", solutionState.Spec.RootResource, namespace)
+					_, err := i.ApiClient.GetSolutionContainer(ctx, solutionState.Spec.RootResource, namespace, i.Config.User, i.Config.Password)
+					if err != nil && strings.Contains(err.Error(), constants.NotFound) {
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:226]: NotFound is an error we can handle. So you can modify the log statement to make it more clear, "failed to get xxxx with error NotFound, try to create ...". How about the error we cannot handle, do you return the error immediately?

File: api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+				Body:  []byte(err.Error()),
+			})
+		}
+		jData, _ := utils.FormatObject(state, isArray, request.Parameters["path"], request.Parameters["doc-type"])
+		resp := observ_utils.CloseSpanWithCOAResponse(span, v1alpha2.COAResponse{
+			State:       v1alpha2.OK,
+			Body:        jData,
+			ContentType: "application/json",
+		})
+		if request.Parameters["doc-type"] == "yaml" {
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go:115]: Why we need this?

File: api/symphony-api.json [hunk 3]
```diff
@@ -13,17 +13,23 @@
       "lng": "-122.12826"
     },
     "currentSite": {
-      "baseUrl": "http://symphony-service:8080/v1alpha2/",
+      "baseUrl": "https://symphony-service:8081/v1alpha2/",
```
Comment (msftcoderdjw) [inline api/symphony-api.json:16]: I don't think we can run standalone mode with https. We didn't bind 8081 port and we don't have a cert generator.

File: api/symphony-api.json [hunk 4]
```diff
@@ -13,17 +13,23 @@
...
-    }
+    }  
   },
   "api": {
     "pubsub": {
       "shared": true,
       "provider": {
-        "type": "providers.pubsub.memory",
-        "config": {}
+        "type": "providers.pubsub.redis",
```
Comment (msftcoderdjw) [inline api/symphony-api.json:25]: Why you change this?

File: api/symphony-api.json [hunk 5]
```diff
@@ -56,8 +61,10 @@
           {
             "name": "stage-manager",
             "type": "managers.symphony.stage",
-            "properties": {   
-              "providers.state": "k8s-state"        
+            "properties": {
```
Comment (msftcoderdjw) [inline api/symphony-api.json:64]: Please review your change on this file. I doubt it messed with other files.

File: k8s/apis/fabric/v1/target_webhook.go [hunk 6]
```diff
@@ -81,6 +82,25 @@ func (r *Target) Default() {
 	if r.Spec.ReconciliationPolicy != nil && r.Spec.ReconciliationPolicy.State == "" {
 		r.Spec.ReconciliationPolicy.State = v1.ReconciliationPolicy_Active
 	}
+
+	if r.Spec.RootResource != "" {
+		var targetContainer TargetContainer
+		err := myTargetClient.Get(context.Background(), client.ObjectKey{Name: r.Spec.RootResource, Namespace: r.Namespace}, &targetContainer)
```
Comment (msftcoderdjw) [inline k8s/apis/fabric/v1/target_webhook.go:88]: Do you add indexfield on root resource?

File: packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml [hunk 7]
```diff
@@ -2705,6 +2705,26 @@ webhooks:
     resources:
     - solutions
   sideEffects: None
+- admissionReviewVersions:
```
Comment (msftcoderdjw) [inline packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml:2708]: Did you try mage helmTemplate to rebuild this file?

**Review by msftcoderdjw — CHANGES_REQUESTED** [review]

File: api/symphony-api.json (hunk 4 above)
Comment (LeiQL) [reply api/symphony-api.json:25]: This file is mis-replaced. will recover back

File: packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml (hunk 7 above)
Comment (LeiQL) [reply packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml:2708]: Yes. Any concerns?

Comment (msftcoderdjw) [reply packages/helm/symphony/templates/symphony-core/symphonyk8s.yaml:2708]: No concerns. We need to make sure symphonyk8s.yaml can be auto-generated.

File: k8s/apis/fabric/v1/target_webhook.go (hunk 6 above)
Comment (msftcoderdjw) [reply k8s/apis/fabric/v1/target_webhook.go:88]: Checked with @LeiQL , we don't do list by rootResource in informer so adding indexfield can be optional.

File: api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+		"method": "DeleteState",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	err = t.StateProvider.Delete(ctx, states.DeleteRequest{
+		ID: name,
+		Metadata: map[string]interface{}{
+			"namespace": namespace,
+			"group":     model.SolutionGroup,
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go:58]: Wrong group

File: api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	ctx, span := observability.StartSpan("CampaignContainersManager", ctx, &map[string]string{
+		"method": "ListState",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	listRequest := states.ListRequest{
+		Metadata: map[string]interface{}{
+			"version":   "v1",
+			"group":     model.SolutionGroup,
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go:117]: wrong group

File: api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go (hunk 8 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go:58]: Good catch. campaign is the only resource which is not supported in materialize.

File: api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go (hunk 9 above)
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go]: Wrong group. You can define the group, resource, name, kind as property in manager, then we can keep them consistent.

File: api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+		"method": "GetSpec",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	getRequest := states.GetRequest{
+		ID: id,
+		Metadata: map[string]interface{}{
+			"version":   "v1",
+			"group":     model.SolutionGroup,
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go]: Wrong group.

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go (hunk 1 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:226]: No. Just to do upsert and webhook will throw out the container doesn't exist. Do you think we need to throw here if not NotFound for pre-handling.

File: api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go (hunk 2 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go:115]: Request contract for all vendors. @Haishi2016 do we need to remove this in container vendors?

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go (hunk 1 above)
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:226]: As a client, if you are trying to check the parent at first and create if not exists. Why we proceed the child resource creation on any errors other than NotFound? 
This step is ensuring the parent resource's existence. If we get an error during the check, we need to stop the flow and fail as earlier as possible. Or you can add retry to handle the transient error.

Comment (LeiQL) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:229]: Updated to throw if not `NotFound`.

File: api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+func getCampaignContainerState(body interface{}, etag string) (model.CampaignContainerState, error) {
+	var CampaignContainerState model.CampaignContainerState
+	bytes, _ := json.Marshal(body)
+	err := json.Unmarshal(bytes, &CampaignContainerState)
+	if err != nil {
+		return model.CampaignContainerState{}, err
+	}
+	if CampaignContainerState.Spec == nil {
+		CampaignContainerState.Spec = &model.CampaignContainerSpec{}
+	}
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go:149]: I would recommend initializing nil Status field as well for all the containerState

File: api/pkg/apis/v1alpha1/vendors/campaigns-container-vendor.go [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+				Body:  []byte(err.Error()),
+			})
+		}
+
+		return observ_utils.CloseSpanWithCOAResponse(span, v1alpha2.COAResponse{
+			State: v1alpha2.OK,
+		})
+	case fasthttp.MethodDelete:
+		ctx, span := observability.StartSpan("onCampaignContainers-DELETE", pCtx, nil)
+		err := c.CampaignContainersManager.DeleteState(ctx, id, namespace)
```
Comment (Haishi2016) [inline api/pkg/apis/v1alpha1/vendors/campaigns-container-vendor.go:143]: What happens to the campaigns under the container? Maybe to use a Campaign manager to check children before deletion is allowed?

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go [hunk 13]
```diff
@@ -159,11 +160,39 @@ func (i *MaterializeStageProvider) Process(ctx context.Context, mgrContext conte
...
+					if len(parts) == 2 {
+						instanceState.Spec.RootResource = parts[0]
+						instanceState.Spec.Version = parts[1]
+					} else {
+						mLog.Errorf("Instance name is invalid: instance - %s, catalog - %s", instanceName, name)
+						return outputs, false, v1alpha2.NewCOAError(nil, fmt.Sprintf("Instance name is invalid: catalog - %s", name), v1alpha2.BadRequest)
+					}
+
+					mLog.Debugf("  P (Materialize Processor): check instance contains %v, namespace %s", instanceState.ObjectMeta.Name, namespace)
+					_, err := i.ApiClient.GetInstanceContainer(ctx, instanceState.Spec.RootResource, namespace, i.Config.User, i.Config.Password)
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:180]: You need to fix create stage provider as well.

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go [hunk 14]
```diff
@@ -159,11 +160,39 @@ func (i *MaterializeStageProvider) Process(ctx context.Context, mgrContext conte
...
+					if err != nil && strings.Contains(err.Error(), constants.NotFound) {
```
Comment (RemindD) [inline api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:181]: use v1alpha2.IsNotFound to check 404 response

Comment (LeiQL) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:181]: No I think. APIclient handle notfound as `resp.StatusCode == 404 { // API service is already gone`

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go (hunk 13 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:180]: Create stage is to create object. If version object is created by the stage and container object doesn't exist, to fail the request is better than handling container creation in `create ` stage, not the same case for create.

File: api/pkg/apis/v1alpha1/vendors/campaigns-container-vendor.go (hunk 12 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/vendors/campaigns-container-vendor.go:143]: Maybe we can register container webhooks to guard deletion if children count is not empty. Then the behavior via AMR or kubectl in local system will be aligned. Then back to consider API users, if we would guard the same, need to update manager to block deletion but API users behavior will not block any cases right now. So we can leave container manager clean without children check? What do you think?
TODO will be:
- Add container webhooks
- Validate deletion: check children whose ownerReference has the current container uid.
- Leave container the same.

Comment (LeiQL) [reply api/pkg/apis/v1alpha1/vendors/campaigns-container-vendor.go:143]: Updated in the latest commit. Container webhook added to validate children resources for deletion.

File: k8s/apis/solution/v1/solution_webhook.go [hunk 15]
```diff
@@ -33,6 +33,11 @@ func (r *Solution) SetupWebhookWithManager(mgr ctrl.Manager) error {
 		solution := rawObj.(*Solution)
 		return []string{solution.Spec.DisplayName}
 	})
+	mgr.GetFieldIndexer().IndexField(context.Background(), &Solution{}, ".spec.rootResource", func(rawObj client.Object) []string {
```
Comment (msftcoderdjw) [inline k8s/apis/solution/v1/solution_webhook.go:36]: Is there any scenario that users only configure ownerReference but forget spec.rootResource?

File: k8s/apis/fabric/v1/targetcontainer_webhook.go [hunk 16]
```diff
@@ -0,0 +1,122 @@
+/*
```
Comment (msftcoderdjw) [inline k8s/apis/fabric/v1/targetcontainer_webhook.go:1]: Given you added more webhooks, consider to add more tests cases in following PR.

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go (hunk 14 above)
Comment (RemindD) [inline api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:181]: I don't think we should have different behaviors for different providers. If materialize provider creates containers if containers don't exist, create provider should also do the same way.

Comment (RemindD) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:181]: This is not correct. The statusCode of not found is not 404 for api server.

File: api/constants/constants.go [hunk 17]
```diff
@@ -24,6 +24,7 @@ const (
 	GroupPrefix     = "symphony"
 	ManagerMetaKey  = GroupPrefix + "/managed-by"
 	InstanceMetaKey = GroupPrefix + "/instance"
+	NotFound        = "Not Found"
```
Comment (RemindD) [inline api/constants/constants.go:27]: use v1alpha2.IsNotFound instead. error type comparison is better than error context comparison.

Comment (LeiQL) [reply api/constants/constants.go:27]: Not correct. Inside MVP you can always use `v1alpha2.IsNotFound` as the error type is internal. But this error is coverted from API call which is a typical genetic error, cannot be converted to v1alpha.error. Sample log: `level=error msg="V (InstanceContainers): onInstanceContainers failed - Not Found: failed to get object (caused by: instancecontainers.solution.symphony \"siteinstance\" not found)`

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go (hunk 14 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:181]: Sync offline. Users should create container objects before using create stage to create a versioned object.

File: k8s/apis/solution/v1/solution_webhook.go (hunk 15 above)
Comment (LeiQL) [reply k8s/apis/solution/v1/solution_webhook.go:36]: No such scenario. `rootResource` is a required filed and checked in webhook.

File: api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go (hunk 14 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go:181]: Double verified. APIclient error cannot be used within function ` v1alpha2.IsNotFound()`

File: api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go (hunk 2 above)
Comment (Haishi2016) [reply api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go:115]: We can keep these. These are from earlier customer feedbacks where they want to consume query result in either JSON or YAML and they want to be able to do simple extractions from the query results

File: api/constants/constants.go (hunk 17 above)
Comment (msftcoderdjw) [reply api/constants/constants.go:27]: String match is not reliable. Maybe in future, we need to consider if we can populate the inner error and keep the original type then do comparison.

**Review by msftcoderdjw — APPROVED** [review]
Approved this PR for feature release. But we still need to do some follow ups, including,
1. Add more test cases for container objects and webhooks check. And also provide docs and samples.
2. Consider on how to reuse the code for container objects.
3. Improve error return in api client.
Let's open some work items.

File: api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go (hunk 2 above)
Comment (LeiQL) [reply api/pkg/apis/v1alpha1/vendors/targets-container-vendor.go:115]: Thanks @Haishi2016. Will resolve this conversation.

**Review by Haishi2016 — APPROVED** [review]
I'm okay with the PR, but please make sure TODOs are captured as issues. We need to keep consistency between standalone and k8s mode. I just came from a customer meeting who explicitly requested to run outside of K8s.

**Pull request merged**